File: src/main/java/org/apache/prepbuddy/normalizers/ZScoreNormalizer.java
package org.apache.prepbuddy.normalizers;

import org.apache.prepbuddy.rdds.TransformableRDD;
import org.apache.spark.api.java.JavaDoubleRDD;

/**
 * A normalizer technique which normalizes data by their standard score.
 * Formula for Z Score Normalization : (X - Mean) / Standard Deviation.
 */
public class ZScoreNormalizer implements NormalizationStrategy {

    private Double standardDeviation;
    private Double mean;

    @Override
    public void prepare(TransformableRDD transformableRDD, int columnIndex) {
        JavaDoubleRDD doubleRDD = transformableRDD.toDoubleRDD(columnIndex);
        standardDeviation = doubleRDD.stdev();
        mean = doubleRDD.mean();
    }

    @Override
    public String normalize(String rawValue) {
        return String.valueOf((Double.valueOf(rawValue) - mean) / standardDeviation);
    }
}

File: src/main/scala/org/apache/datacommons/prepbuddy/normalizers/ZScoreNormalizer.scala
package org.apache.datacommons.prepbuddy.normalizers

import org.apache.datacommons.prepbuddy.rdds.TransformableRDD

/**
  * A normalizer technique which normalizes data by their standard score.
  * Formula for Z Score Normalization : (X - Mean) / Standard Deviation.
  */
class ZScoreNormalizer extends NormalizationStrategy {

    private var standardDeviation: Double = 0

    private var mean: Double = 0

    override def prepare(transformableRDD: TransformableRDD, columnIndex: Int): Unit = {
        val doubleRDD = transformableRDD.toDoubleRDD(columnIndex)
        standardDeviation = doubleRDD.stdev
        mean = doubleRDD.mean
    }

    override def normalize(rawValue: String): String = {
        String.valueOf((rawValue.toDouble - mean) / standardDeviation)
    }
}
